module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+			    FIEMAP_EXTENT_DELALLOC | FIEMAP_EXTENT_UNKNOWN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, start, size, 0, 0, 0,
+		    FIEMAP_EXTENT_DELALLOC | FIEMAP_EXTENT_UNKNOWN |
+		    FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * Pending free extents set FIEMAP_EXTENT_UNWRITTEN since they will be a hole.
+ * FIEMAP_EXTENT_DELALLOC is set to indicate it has not yet been written.  The
+ * FIEMAP_EXTENT_UNKNOWN flag must be set when FIEMAP_EXTENT_DELALLOC is set.
+ * Free extents are only inserted in to the first extent tree.
+ */
+static void
+zfs_fiemap_add_free(void *arg, uint64_t start, uint64_t size)
+{
+	zfs_fiemap_t *fm = (zfs_fiemap_t *)arg;
+	avl_tree_t *t = &fm->fm_extent_trees[0];
+
+	zfs_fiemap_clear(t, start, size);
+
+	if (fm->fm_flags & FIEMAP_FLAG_NOMERGE) {
+		uint64_t blksz = fm->fm_block_size;
+
+		for (uint64_t i = start; i < start + size; i += blksz) {
+			zfs_fiemap_add_impl(t, i, blksz, 0, 0, 0,
+			    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_DELALLOC |
+			    FIEMAP_EXTENT_UNKNOWN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, start, size, 0, 0, 0,
+		    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_DELALLOC |
+		    FIEMAP_EXTENT_UNKNOWN | FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * The entire file is sparse and there are no level zero blocks with data.
+ * In this case pretend that hole block pointers exist to maintain consistency
+ * in the reported output.  Either add a single unwritten extent for the
+ * entire length of the file.  Or when no merging is requested add the
+ * correct number of hole block pointers.  Only the first extent tree should
+ * be populated since only holes are being added.
+ */
+static void
+zfs_fiemap_add_sparse(zfs_fiemap_t *fm)
+{
+	avl_tree_t *t = &fm->fm_extent_trees[0];
+	uint64_t blksz = fm->fm_block_size;
+	uint64_t size = P2ROUNDUP(fm->fm_file_size, blksz);
+
+	if (fm->fm_flags & FIEMAP_FLAG_NOMERGE) {
+		for (uint64_t i = 0; i < size; i += blksz) {
+			zfs_fiemap_add_impl(t, i, blksz, 0, 0, 0,
+			    FIEMAP_EXTENT_UNWRITTEN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, 0, size, 0, 0, 0,
+		    size == blksz ? FIEMAP_EXTENT_UNWRITTEN :
+		    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * Walk the block pointers for the provided object and assemble a tree
+ * of extents which describe the logical to physical mapping.  Additionally
+ * include dirty buffers for the object which will be written but have
+ * not yet have had space allocated on disk.
+ */
+int
+zfs_fiemap_assemble(struct inode *ip, zfs_fiemap_t *fm)
+{
+	znode_t *zp = ITOZ(ip);
+	zfsvfs_t *zfsvfs = ZTOZSB(zp);
+	zbookmark_phys_t czb;
+	txg_handle_t th;
+	dnode_t *dn;
+	spa_t *spa;
+	uint64_t open_txg, syncing_txg, dirty_txg;
+	int error;
+
+	ZFS_ENTER(zfsvfs);
+	ZFS_VERIFY_ZP(zp);
+
+	error = dnode_hold(zfsvfs->z_os, zp->z_id, FTAG, &dn);
+	if (error) {
+		ZFS_EXIT(zfsvfs);
+		return (error);
+	}
+
+	spa = dmu_objset_spa(dn->dn_objset);
+
+	if (fm->fm_flags & FIEMAP_FLAG_SYNC)
+		txg_wait_synced(spa_get_dsl(spa), 0);
```

Comment:
Let's do this only if it could change the result (i.e. the dnode is dirty).

Reply:
Agreed.  Otherwise, the `ioctl(FIEMAP)` call will be blocked on TXG sync repeatedly and be serial calls will be rate-limited to 1/txg_interval?

Reply:
Will fix